### PR TITLE
Add Deploy Scripts

### PR DESCRIPTION
**Pull request by Psirex**

### CODE REVIEW

**Review by bulbozaur — APPROVED** [review]
LGTM

**Review by avsetsin — APPROVED** [review]
LGTM

Consider description of how to run the deployment scripts in the readme

File: hardhat.config.ts [hunk 1]
```diff
@@ -5,30 +5,42 @@ import './utils/assert'
 
 dotenv.config()
 
+const MAINNET_RPC_URL = process.env.RPC_URL
+const GOERLI_RPC_URL = process.env.GOERLI_RPC_URL
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply hardhat.config.ts:9]: +

File: scripts/create-amount-converter-instance.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+import fmt from '../utils/format'
+import { confirmOrAbort } from '../utils/prompt'
+import { getDeployer, verify } from '../utils/deployment'
+import { AmountConverterFactory__factory } from '../typechain-types'
+import { AmountConverterDeployedEvent } from '../typechain-types/contracts/factories/AmountConverterFactory'
+
+const AMOUNT_CONVERTER_FACTORY = ''
+
+const CONVERSION_TARGET = ''
```
Comment (TheDZhon) [inline scripts/create-amount-converter-instance.ts:12]: the name is quite vague

File: scripts/create-amount-converter-instance.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  // prettier-ignore
+  console.log(
+      `Preparing for ${fmt.name('AmountConverter')} deployment on "${fmt.network(network.name)}" network,`,
+      `using the ${fmt.name("AmountConverterFactory")} on address ${fmt.address(AMOUNT_CONVERTER_FACTORY)} ...\n`
+  )
+
+  const deployer = await getDeployer()
+
+  console.log(`Deployment parameters:`)
+  console.log(`  * Conversion target: ${fmt.value(CONVERSION_TARGET)}`)
```
Comment (TheDZhon) [inline scripts/create-amount-converter-instance.ts:37]: still hard to catch what does it mean: 'conversion target'

File: scripts/create-amount-converter-instance.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  console.log(
+    `  * Allowed tokens to sell: ${fmt.value('[' + ALLOWED_TOKENS_TO_SELL.join(', ') + ']')}`
+  )
+  console.log(
+    `  * Allowed tokens to buy: ${fmt.value('[' + ALLOWED_TOKENS_TO_BUY.join(', ') + ']')}`
+  )
+  console.log(
+    `  * Price feeds heartbeat timeouts: ${fmt.value(
```
Comment (TheDZhon) [inline scripts/create-amount-converter-instance.ts:45]: Would be easier to read if the values were mapped with the 'sell' items (using the table, e.g.)

File: scripts/create-amount-converter-instance.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  if (network.name !== 'hardhat') {
+    await verify(amountConverterAddress, [
+      CONVERSION_TARGET,
+      ALLOWED_TOKENS_TO_SELL,
+      ALLOWED_TOKENS_TO_BUY,
+      PRICE_FEEDS_HEARTBEAT_TIMEOUTS,
+    ])
+  } else {
+    console.log(`Deploying on the test network, verification is skipped.`)
```
Comment (TheDZhon) [inline scripts/create-amount-converter-instance.ts:94]:
```suggestion
    console.log(`Deployed on the local hardhat network, verification is skipped.`)
```

File: scripts/create-stonks-instances.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+  const deployer = await getDeployer()
+
+  for (const [pair, config] of Object.entries(STONKS_CONFIGS)) {
+    console.log(`Deploy config for pair ${fmt.name(pair)}:`)
+    console.log(`  * token from: ${fmt.value(config.tokenFrom)}`)
+    console.log(`  * token to: ${fmt.value(config.tokenTo)}`)
+    console.log(`  * order duration (sec): ${fmt.value(config.orderDurationInSeconds)}`)
+    console.log(`  * margin (bps): ${fmt.value(config.marginBasisPoints)}`)
+    console.log(`  * price tolerance (bps): ${fmt.value(config.priceToleranceInBasisPoints)}`)
```
Comment (TheDZhon) [inline scripts/create-stonks-instances.ts:39]: Those names are pretty hard to memorize (i.e., which one is the upper bound or lower), I would suggest disambiguating them a bit with additional printable hints

File: scripts/create-stonks-instances.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+async function main() {
+  // prettier-ignore
+  console.log(
+    `Preparing for ${fmt.name('Stonks')} contracts deployment on "${fmt.network(network.name)}" network,`,
+    `using the ${fmt.name("StonksFactory")} on address ${fmt.address(STONKS_FACTORY)} ...\n`
+  )
+
+  const deployer = await getDeployer()
+
+  for (const [pair, config] of Object.entries(STONKS_CONFIGS)) {
```
Comment (TheDZhon) [inline scripts/create-stonks-instances.ts:33]: shall we output the total number of configs?

File: scripts/create-stonks-instances.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    console.log(`Deploy config for pair ${fmt.name(pair)}:`)
```
Comment (TheDZhon) [inline scripts/create-stonks-instances.ts:34]: would suggest adding an output of the config index (as below)

File: scripts/create-stonks-instances.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    console.log(
+      [
+        `The ${fmt.name('Stonks')} instance for the pair ${fmt.name(pair)}`,
+        `was deployed successfully: ${fmt.address(stonksAddress)}\n`,
+      ].join(' ')
+    )
+    if (network.name !== 'hardhat') {
+      await verify(stonksAddress, deployArgs as unknown as unknown[])
+    } else {
+      console.log(`Deploying on the test network, verification is skipped.`)
```
Comment (TheDZhon) [inline scripts/create-stonks-instances.ts:96]:
```suggestion
      console.log(`Deployed on the local hardhat network, verification is skipped.`)
```

File: scripts/deploy-amount-converter-factory.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  const address = await amountConverter.getAddress()
+  console.log(
+    `The ${fmt.name('AmountConverterFactory')} was deployed successfully: ${fmt.address(address)}`
+  )
+
+  if (network.name !== 'hardhat') {
+    await verify(address, [CHAINLINK_PRICE_FEED_REGISTRY])
+  } else {
+    console.log(`Deploying on the test network, verification is skipped.`)
```
Comment (TheDZhon) [inline scripts/deploy-amount-converter-factory.ts:40]:
```suggestion
    console.log(`Deployed on the local hardhat network, verification is skipped.`)
```

File: scripts/deploy-stonks-factory.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+import fmt from '../utils/format'
+import { confirmOrAbort } from '../utils/prompt'
+import { StonksFactory__factory } from '../typechain-types'
+import { getDeployer, verify } from '../utils/deployment'
+import { OrderSampleDeployedEvent } from '../typechain-types/contracts/factories/StonksFactory'
+
+const AGENT = ''
+const SETTLEMENT = ''
+const VAULT_RELAYER = ''
```
Comment (TheDZhon) [inline scripts/deploy-stonks-factory.ts:11]: Those names are a bit generic to remember their meaning

File: scripts/deploy-stonks-factory.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    `The ${fmt.name('StonksFactory')} contract was deployed successfully: ${fmt.address(stonksAddress)}\n`
+  )
+  console.log(
+    `Sample of the ${fmt.name('Order')} contract was deployed at ${fmt.address(orderAddress)}\n`
+  )
+  if (network.name !== 'hardhat') {
+    await verify(stonksAddress, [AGENT, SETTLEMENT, VAULT_RELAYER])
+    await verify(orderAddress, [AGENT, SETTLEMENT, VAULT_RELAYER])
+  } else {
+    console.log(`Deploying on the test network, verification is skipped.`)
```
Comment (TheDZhon) [inline scripts/deploy-stonks-factory.ts:70]:
```suggestion
    console.log(`Deployed on the local hardhat network, verification is skipped.`)
```

File: scripts/deploy-stonks-factory.ts (hunk 11 above)
Comment (TheDZhon) [reply scripts/deploy-stonks-factory.ts:11]: maybe:
`COWSWAP_SETTLEMENT` and `COWSWAP_VAULT_RELAYER`? (following the same pattern for `CHAINLINK_PRICE_FEED_REGISTRY`)

**Review by TheDZhon — APPROVED** [review]
LGTM, though please consider some suggestions

File: hardhat.config.ts [hunk 13]
```diff
@@ -5,30 +5,42 @@ import './utils/assert'
...
+
+if (!MAINNET_RPC_URL && !GOERLI_RPC_URL) {
+  throw new Error(`RPC url was not provided. Please, ensure the .env file is filled correctly.`)
+}
+
 const WALLET_PRIVATE_KEY = process.env.WALLET_PRIVATE_KEY!
```
Comment (TheDZhon) [inline hardhat.config.ts:15]: :grey_question: maybe have `MAINNET_WALLET_PRIVATE_KEY` and `GOERLI_WALLET_PRIVATE_KEY` to discourage using the same PK on real and test networks (not a strong opinion, though, posting separately)

File: hardhat.config.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkolpakov) [reply hardhat.config.ts:9]: No needed to support goerli anymore

**Pull request merged**